pyface/data_view/abstract_value_type.py

```diff
@@ -0,0 +1,182 @@
+
+        Parameters
+        ----------
+        model : AbstractDataModel
+            The data model holding the data.
+        row : sequence of int
+            The row in the data model being queried.
+        column : sequence of int
+            The column in the data model being queried.
+        value : any
+            The value to set.
+
+        Returns
+        -------
+        success : bool
+            Whether or not the value was successfully set.
+        """
+        if not self.can_edit(model, row, column):
+            return False
+        return model.set_value(row, column, value)
+
+    def has_text(self, model, row, column):
+        """ Whether or not the value has a textual representation.
+
+        The default implementation returns True if ``get_text``
+        returns a non-empty value.
+
+        Parameters
+        ----------
+        model : AbstractDataModel
+            The data model holding the data.
+        row : sequence of int
+            The row in the data model being queried.
+        column : sequence of int
+            The column in the data model being queried.
+
+        Returns
+        -------
+        has_text : bool
+            Whether or not the value has a textual representation.
+        """
+        return self.get_text(model, row, column) != ""
+
+    def get_text(self, model, row, column):
+        """ The textual representation of the underlying value.
+
+        The default implementation calls str() on the underlying value.
+
+        Parameters
+        ----------
+        model : AbstractDataModel
+            The data model holding the data.
+        row : sequence of int
+            The row in the data model being queried.
+        column : sequence of int
+            The column in the data model being queried.
+
+        Returns
+        -------
+        text : str
+            The value to edit.
+        """
+        return str(model.get_value(row, column))
+
+    def set_text(self, model, row, column, text):
+        """ Set the text of the underlying value.
+
+        This is provided primarily for backends which may not permit
+        non-text editing of values, in which case this provides an
+        alternative route to setting the value.  The default implementation
+        does not allow setting the text.
+
+        Parameters
+        ----------
+        model : AbstractDataModel
+            The data model holding the data.
+        row : sequence of int
+            The row in the data model being queried.
+        column : sequence of int
+            The column in the data model being queried.
+        text : str
+            The text to set.
+
+        Returns
+        -------
+        success : bool
+            Whether or not the value was successfully set.
+        """
+        return False
+
+    @observe('+update')
+    def update_value_type(self, event=None):
+        """ Fire update event when marked traits change. """
+        self.updated = True
```

Comment:
The `update` metadata is part of the documented public API of the base class; classes which inherit from it should be aware of the behaviour.  If you don't want the "update" behaviour, it is an `abc`, so write your own and `register` it.

Issues about GUI listeners and the 'update' metadata are moot, because they are all listening to the `AbstractDataModel` API events, which should be hooked up and unhooked up dynamically to avoid the sorts of problems you are worried about; the `AbstractDataModel` subclasses are responsible for listening to the `update` event of their value types and translating those to the appropriate events they they should fire (eg. look at the `ArrayDataModel`).

Perhaps "update" is too generic, however.  There may be a better name to use for the metadata to make it easier to re-use trait types.

Reply:
Thank you. I am not too concerned about the name `"update"` so I will leave that up to you.

There is another concern... suppose a trait is changed and the model needs to do something about that change *before* the `updated` event is fired. The correct way is that this `update_value_type` needs to be overridden.  But it is conceivable that a developer might just *add* a change handler and it might happen to work, but that would be unsafe because the order of which change handlers are called is not guaranteed: It depends on when the change handlers are registered in HasTraits's meta class.

For example:
```

class MyValueType(AbstractValueType):
    my_trait = Int(update=True)

    readonly = Bool()

    def can_edit(self, model, row, column):
        return not self.readonly

    @observe("my_trait")
    def my_trait_updated(self, event):
        self.readonly = event.new > 10

```

They might be relying on the `update_value_type` being called after `my_trait_updated` is called. Nothing in traits guarantee that, and I think we want to discourage that pattern and force developers to write this:
```
class MyValueType(AbstractValueType):
    # assume the update_value_type is removed.

    my_trait = Int(update=True)

    readonly = Bool()

    def can_edit(self, model, row, column):
        return not self.readonly

    @observe("my_trait")
    def my_trait_updated(self, event):
        self.readonly = event.new > 10
        self.updated = True
```

Reply:
I think that the first isn't a problem, because all trait handling will be over and done with before the Qt event loop gets to do anything more, so `can_edit` won't get called until _after_ both handlers are done (this is assuming changes happen on the main thread, which is a good enough assumption for now).

Reply:
No, it will likely be part of the public API (if only so that people can specify `Instance(...)` for it), and people will write subclasses for it, but I expect that they will do so rarely (because we will quickly build a set of classes that cover 90% of use cases) and will likely subclass from intermediate helper classes more than directly from the base class.  It's not complete yet, and so it's not the focus of this PR, and will likely eventually be in the .api files.

I expect _virtual_ subclasses to be extraordinarily rare.

As I said, I think this class is going to be _very_ stable once complete, and if, in the future, we decide that we need to add to it Python has more than enough introspective features to allow us to manage the transition (eg. do a `hasattr` check before calling a new method and providing a sane default and deprecation warning if it isn't present).

Reply:
That's good to know. Thanks!